### PR TITLE
add joins, insertion & search to correlation engine

### PR DESCRIPTION
**Pull request by sbcd90**
### Description
This PR focusses on the `Join` handler, `Insertion` handler & the `Search` handler components of the Events Correlation Engine. The component level architecture diagram can be found [here](https://user-images.githubusercontent.com/617607/226722234-e78e338f-0eec-4440-b480-c8049780f963.png)

## Join Handler
The Join handler component determines immediate neighbors of a particular event using the Correlation Rules defined by the user for the log indices(or datastreams) they wish to correlate.

Here is a brief schematic class diagram of the Join Handler layer.
image

## Insertion Handler
In the Insertion handler layer events are converted to k-dimensional vectors & are stored in the vector storage layer mentioned above along with their correlations.

Here is a brief schematic class diagram of the Insertion Handler layer.
image

## Search Handler
The Search Handler layer allows user to specify a particular event, & then converts it to a k-dimensional vector & then uses it to query its neighboring eventswhich are actually its correlated events within a time window.

Here is a brief schematic class diagram of the Search Handler layer.
image

### REST apis

The following are the REST apis introduced by this PR.

#### Correlate an event

```
POST /_correlation/events

Request:
{
  "index": "app_logs",
  "event": "l2UZD4kBSz-dGVL1EZFJ",
  "store": false
}

Response:
{
  "is_orphan": false,
  "neighbor_events": {
    "windows": [
      "DmyTDI8B1GiWmmBKbkSu"
    ]
  }
}
```

#### Search Correlated Events for the input event

```
GET /_correlation/events?index=windows&event=FZ-9DI8BfF4uBOcpmLhq&timestamp_field=winlog.timestamp&time_window=300000&nearby_events=5

Response:
{
  "events": [
    {
      "index": "app_logs",
      "event": "Fp-9DI8BfF4uBOcpmLj1",
      "score": 0.5,
      "tags": []
    }
  ]
}
```

## 
### Related Issues
#6854

### Check List
- [X] New functionality includes testing.
  - [X] All tests pass
- [] New functionality has been documented.
  - [X] New functionality has javadoc added
- [X] Commits are signed per the DCO using --signoff
- [X] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by getsaurabh02** [comment]
@andrross Looks like the PR got closed as it was taking some time for the review to complete. I am taking first pass this week and would request you to then take a look as well. Can we re-open this please.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 1]
```diff
@@ -0,0 +1,498 @@
...
+            searchRequest.source(searchSourceBuilder);
+
+            client.search(searchRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    if (response.isTimedOut()) {
+                        onFailures(new OpenSearchStatusException(response.toString(), RestStatus.REQUEST_TIMEOUT));
+                    }
+
+                    Iterator<SearchHit> hits = response.getHits().iterator();
```
Comment (shwetathareja) [inline plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:147]: is there any limits on the no. of hits?

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    onFailures(e);
+                }
+            });
+        }
+
+        private void prepRulesForCorrelatedEventsGeneration(String index, String event, List<CorrelationRule> correlationRules) {
```
Comment (shwetathareja) [inline plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:173]: there are very long method, can break into multiple small private methods

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * Transport Action for indexing correlations for a particular event
+ *
+ * @opensearch.internal
+ */
+public class TransportIndexCorrelationAction extends HandledTransportAction<IndexCorrelationRequest, IndexCorrelationResponse> {
```
[the remark this answers is not in the document]
Comment (shwetathareja) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:64]: Applicable for all transport classes

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+        this.correlationTimeWindow = EventsCorrelationSettings.CORRELATION_TIME_WINDOW.get(this.settings).getMillis();
+
+        this.clusterService.getClusterSettings()
+            .addSettingsUpdateConsumer(EventsCorrelationSettings.CORRELATION_TIME_WINDOW, it -> correlationTimeWindow = it.getMillis());
+    }
+
+    @Override
+    protected void doExecute(Task task, IndexCorrelationRequest request, ActionListener<IndexCorrelationResponse> listener) {
+        AsyncIndexCorrelationAction asyncAction = new AsyncIndexCorrelationAction(request, listener);
+        asyncAction.start();
```
Comment (shwetathareja) [inline plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:110]: These are heavy weight transport actions chaining multiple queries as well. In which threadpool they would run. They shouldn't run on transport threadpool.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportSearchCorrelatedEventsAction.java [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *  Transport Action for searching correlated events of a particular event
+ *
+ * @opensearch.internal
+ */
+public class TransportSearchCorrelatedEventsAction extends HandledTransportAction<
```
Comment (shwetathareja) [inline plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportSearchCorrelatedEventsAction.java:44]: Are there any stats planned for CorrelatedEvents Actions?

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/resthandler/RestIndexCorrelationAction.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import org.opensearch.rest.action.RestResponseListener;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+
+/**
+ * Rest action for indexing an event and its correlations
+ */
+public class RestIndexCorrelationAction extends BaseRestHandler {
```
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/resthandler/RestIndexCorrelationAction.java:34]: added search timeouts to these actions.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 7]
```diff
@@ -0,0 +1,498 @@
...
+            searchSourceBuilder.fetchSource(true);
+
+            SearchRequest searchRequest = new SearchRequest();
+            searchRequest.indices(CorrelationRule.CORRELATION_RULE_INDEX);
+            searchRequest.source(searchSourceBuilder);
+
+            client.search(searchRequest, new ActionListener<>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    if (response.isTimedOut()) {
```
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:151]: added check for 429

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 8]
```diff
@@ -0,0 +1,498 @@
...
+                    @Override
+                    public void onResponse(MultiSearchResponse items) {
+                        MultiSearchResponse.Item[] responses = items.getResponses();
+                        Map<String, List<CorrelationQuery>> indexQueriesMap = new HashMap<>();
+                        Long timestamp = null;
+
+                        int idx = 0;
+                        for (MultiSearchResponse.Item response : responses) {
+                            if (response.isFailure()) {
+                                log.error("error:", response.getFailure());
```
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java]: fixed this.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:66]: can address this in a follow-up pr.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 9]
```diff
@@ -0,0 +1,498 @@
...
+                        Map.of(),
+                        List.of()
+                    );
+                    client.execute(StoreCorrelationAction.INSTANCE, storeCorrelationRequest, new ActionListener<>() {
+                        @Override
+                        public void onResponse(StoreCorrelationResponse response) {
+                            if (response.getStatus().equals(RestStatus.OK)) {
+                                onOperation(true, new HashMap<>());
+                            } else {
+                                onFailures(new OpenSearchStatusException("Failed to store correlations", RestStatus.INTERNAL_SERVER_ERROR));
```
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:338]: currently, `RestStatus.OK -> success`, `any other status -> failures`. This is returned by a custom transport action.

[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:338]: i can enhance error handling in a follow-up pr.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 10]
```diff
@@ -0,0 +1,498 @@
...
+            Map<String, List<CorrelationQuery>> indexQueriesMap
+        ) {
+            MultiSearchRequest mSearchRequest = new MultiSearchRequest();
+
+            for (Map.Entry<String, List<CorrelationQuery>> indexQueriesEntry : indexQueriesMap.entrySet()) {
+                String index = indexQueriesEntry.getKey();
+                List<CorrelationQuery> correlationQueries = indexQueriesEntry.getValue();
+
+                // assuming all queries belonging to an index use the same timestamp field.
+                String timestampField = correlationQueries.get(0).getTimestampField();
```
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:366]: added validation.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java [hunk 11]
```diff
@@ -0,0 +1,498 @@
...
+                        MultiSearchResponse.Item[] responses = items.getResponses();
+                        Map<String, Set<String>> eventsAdjacencyList = new HashMap<>();
+
+                        for (MultiSearchResponse.Item response : responses) {
+                            if (response.isFailure()) {
+                                // suppress exception
+                                continue;
+                            }
+
+                            Iterator<SearchHit> searchHits = response.getResponse().getHits().iterator();
```
[the remark this answers is not in the document]
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:429]: yes. pagination is missing. can address this in follow-up prs.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java (hunk 1 above)
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java]: `10000` is default limit.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java (hunk 2 above)
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:190]: can address this refactoring in a follow-up pr.

File: plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java (hunk 4 above)
Comment (sbcd90) [reply plugins/events-correlation-engine/src/main/java/org/opensearch/plugin/correlation/events/transport/TransportIndexCorrelationAction.java:117]: today, all of them run in transport threadpool.

**Review by andrross — CHANGES_REQUESTED** [review]
I do think a correlation engine is useful and believe that the OpenSearch Project should deliver this feature. However, I remain unconvinced that the OpenSearch repository maintainers are the right people to take ownership of this code base. I personally do not have the bandwidth to build up the domain expertise required to properly review this code, and the fact that this PR has languished for over a year suggests that the case is similar for other maintainers. Since this feature is being architected as a plugin, I don't think there is any technical reason this code needs to live inside the OpenSearch repository. I also believe there are significant advantages to being able to iterate and evolve at its own pace, without being coupled to all the bottlenecks and pain points (flaky tests) that exist in the OpenSearch repository. Also note that a [convention has been recently codified](https://github.com/opensearch-project/opensearch-plugins/blob/main/CONVENTIONS.md#creating-new-plugins) that plugins should be developed in their own repository as opposed to inside the OpenSearch repository. If another maintainer wants to make the case that this is the right place for this feature, then I will happily withdraw my objection.

/cc @reta @dblock @shwetathareja

**Comment by dblock** [comment]
I agree with @andrross, should we close this?

**Comment by praveensameneni** [comment]
@dblock, @andrross , @shwetathareja , @reta 
I can understand the reservations on bringing in the plugin into core opensearch repository. As per the [conventions](https://github.com/opensearch-project/opensearch-plugins/blob/main/CONVENTIONS.md) - " .. If you think it should be there anyway, please explain why in the RFC stage." We discussed this topic last year when we first started the RFC and had a [core maintainer opine](https://github.com/opensearch-project/OpenSearch/issues/6779#issuecomment-1485120907) and another at [here](https://github.com/opensearch-project/OpenSearch/issues/6779#issuecomment-1612214109).

I understand maintainers are not permanent and there will be new maintainers and ideas, however as an open source project, what is the guiding principle for these kind of changes. The convention says discuss it during RFC stage and we duly followed the process.

This would be a great precedent for a non-maintainer to contribute and discuss the feature and benefits and qualify to become a core maintainer or we will always be restricting contributions and not grow the project.

**Comment by reta** [comment]
> I understand maintainers are not permanent and there will be new maintainers and ideas, however as an open source project, what is the guiding principle for these kind of changes. The convention says discuss it during RFC stage and we duly followed the process.

@praveensameneni the OpenSearch ecosystem evolves fast and along the way the things we do work or don't become more apparent (plus new challenges always come in). The recently published guidelines (which we did not have a year ago) are trying to touch upon where new plugins belong. I agree with @andrross point and @dblock that it should be out of core at this moment of time. 

 > This would be a great precedent for a non-maintainer to contribute and discuss the feature and benefits and qualify to become a core maintainer or we will always be restricting contributions and not grow the project.

The contributor could become the maintainer on the plugin repo (and continue contribution to core to become a maintainer there as well), we have quite a few successful stories like that already.

**Comment by praveensameneni** [comment]
@reta 
It's currently a plugin inside of core repository, creating an additional plugin repository will also introduce additional dependency and operational overheads. 
Example: Security Analytics has several use cases around correlations, and that is where we started, iterated on and evolved the functionality. Similarly Observability and any log analytics (e.g. application logs) would also have a use case of correlations - show me correlated events when there is an increase in 5xx or response times. And if we follow the separate plugin path, each of the plugins would have to take a dependency on the new plugin (e.g. correlations).

While the guidelines were recently added, the discussion from last year when we created the RFC still holds true.
Also, the recently added guideline says " .._If you think it should be there anyway, please explain why in the RFC stage._"  
It appears the above sentence does not truly mean what it says unless am missing something.

I understand the guidelines around new plugins, which unfortunately is biased towards maintainers domain expertise and does not fully account for integrating dependencies in plugin ecosystem. Additionally, It would also be better to take into account the use case and work backwards from the problem it's trying to solve and the value add for the community.

**Comment by andrross** [comment]
> And if we follow the separate plugin path, each of the plugins would have to take a dependency on the new plugin (e.g. correlations).

@praveensameneni I just want to reiterate that the need to take a dependency on a new plugin is true regardless of where this code lives. The OpenSearch Project [publishes many plugins](https://mvnrepository.com/artifact/org.opensearch.plugin), some of which reside in this repo, some of which reside in other repos, and consumers need not know or care.

> the OpenSearch ecosystem evolves fast and along the way the things that do work or don't become more apparent

What has become apparent to me in this case is that the decision for the OpenSearch maintainers to take ownership of this code has not worked, as shown by this PR being open for over a year. The current state of the correlation engine is that some basic functionality has been committed in this repo on the main branch (I don't think enough is there to be practically useful) and none of it has been backported to 2.x so it has never been released.

**Comment by reta** [comment]
> If it's a separate plugin repo outside of core repo, you take a dependency through common-utils, the plugin repo jar among others, however when shipped as a plugin (e.g. ActionPlugin, ScriptPlugin, ReloadablePlugin, ExtensiblePlugin) inside of core repo, we do not have to take an additional jar, common-utils dependency.

@praveensameneni the plugins are distributed as ZIP archives (bundles), not a JAR files.The OpenSearch distribution comes with some plugins preinstalled / prebundled. What is the plugin repo jar you are referring to?

**Comment by praveensameneni** [comment]
Thank you @reta , @andrross for your insightful comments.
We will create a new plugin repository to support correlations and iterate on the development and release. However, as and when there is more interest in the correlations, we would like to integrate into opensearch-core (not as a plugin) to provide an integrated experience with search and query capabilities leveraging some of the core Lucene constructs.

**Comment by praveensameneni** [comment]
We will take the PR and previously merged PR's into the new opensearch-correlations plugin repository.

**Comment by andrross** [comment]
> If it's a separate plugin repo outside of core repo, you take a dependency through common-utils, the plugin repo jar among others, however when shipped as a plugin (e.g. ActionPlugin, ScriptPlugin, ReloadablePlugin, ExtensiblePlugin) inside of core repo, we do not have to take an additional jar, common-utils dependency.

For posterity, I want to clarify a few points here. There are a couple different types of dependencies between plugins as far as I know.

1. A plugin adds a capability in the core, and another plugin depends on that capability. Example: the k-NN plugin adds vector storage and search capabilities to the existing indexing and search APIs. Plugins like neural-search then build additional features on top of those vector primitives. In this particular case, neural-search also takes a compile dependency on the k-NN jar to get access to some [Java classes defined by k-NN](https://github.com/opensearch-project/neural-search/blob/25d2e82b62b52eae4d73f8f33aab704ff01c0390/src/main/java/org/opensearch/neuralsearch/query/NeuralQueryBuilder.java#L36).
2. A plugin itself defines an extensible capability that other plugins can extend using the Java SPI mechanism. [job-scheduler](https://github.com/opensearch-project/job-scheduler/blob/00d90e82dd5b37cb691484769f957f1586dcdc65/README.md#getting-started) is an example of this.

I'm honestly not sure what category the correlation engine will fall into (thus far I haven't seen it defining an SPI interface so I would guess category 1), but the overall point stands that to a consumer plugin _where_ the dependent plugin code lives should not be important. In both the k-NN and job-scheduler cases they are plugins defined in external repos. I'm not actually familiar with common-utils but I believe it is really just a common library for sharing code across multiple plugins.

---

I'm going to close this PR as it looks like we have a path forward as a separate repository. As the feature matures and the code evolves, if the plugin model is no longer the right fit we can absolutely re-evaluate integrating the feature directly into the core, but for now I believe a separate repository is the right place.

**Comment by andrross** [comment]
I'm going to close this PR as it looks like we have a path forward as a separate repository. As the feature matures and the code evolves, if the plugin model is no longer the right fit we can absolutely re-evaluate integrating the feature directly into the core, but for now I believe a separate repository is the right place.

**Pull request closed without merging**